_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,347 @@
+    in this, we can then update our deployment tooling to lean on this new attribute.
+
+The commits on this PR try to follow RFPR guidelines. Notice how low value and high value changes
+are isolated:
+
+    * Move loop helper code to pkg/loop.
+    * Add ECS dependencies.
+    * Update executive interface mock.
+    * Create an ECSClient interface and generate mock.
+    * Create EcsMetada type to unmarshal ECS metadata.
+    * Create ledger monitor to set ECS 'unhealthy' attribute when falling behind.
+    * Add functional options for the Monitor.
+    * Update reflector config to also compose ledger health config.
+    * Add tests to verify attribute is set correctly using a mocked ECS environment.
+
+Example of what I consider to be a reasonable commit within this PR.
+
+    Create ledger monitor to set ECS 'unhealthy' attribute when falling behind.
+
+    The monitor follows the pattern that the other types use, where a Start(ctx) method blocks, and
+    only exits when the context is canceled.
+
+    It follows this high level logic:
+
+    * Every time the *time.Ticker ticks:
+        * get the ledger latency from the db
+        * compare this value to threshold
+        * if latency is OK, set a 'healthy' attr
+        * else set 'unhealthy' attr
+
+    It keepds the last known successful ECS attr set in a local var so that it doesn't set the attr
+    on each tick. The variable is a *bool because we need three states:
+
+    1. unknown state of ECS attr
+    2. 'healthy'
+    3. 'unhealthy'
+
+
+These are the kinds of PRs I'm thankful to get notified about. I might spend on average 15-30m extra
+to create PRs using these guidelines, but it saves many hours across the reviewers assigned to it,
+and the quality of feedback you get on your RFPR will invariably be higher.
+
+# Tooling
+
+Much of RFPRs ends up being a writing exercise -- writing your PR title, PR description, commit
+titles, commit descriptions, with the reviewer in mind. But a large part of RFPRs involves commit
+discipline. If you're like me, you really only know what the changeset is going to look like after
+you finish making all of your changes. As I approach the finish line, my feature branch has tens if
+not hundreds of garbage commits.
+
+In fact I have a `ci` (checkin) script that enables these garbage commits. I love it:
+
+```bash
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+msg="${*:-wip}"
+git add -A
+git commit -m "${msg}"
+```
+
+I use this constantly and so my feature branch is littered with `wip` commits. These are my
+breadcrumbs that enable to me to "checkpoint" my progress in a way that doesn't take me out of the
+flow. They also let me back out more granular changes -- maybe I wanted to see if a particular change
+made sense. I `ci` before I start, and if I hate it, it's easy for me to throw away that change
+without risk of throwing away what came before it.
```

Comment:
This is a cool idea. I end up relying a lot on undo in my editor which isn't as reliable. I need to get better at using git to checkpoint my work.

Reply:
I think that it shines in particular when the work you're doing is outside of the editor itself (e.g. adding a migration, doing some codegen of some kind, etc). This also makes me realize that I really don't rely on neovim's undo much at all, and I should probably lean into that at some point. Specifically, what does it look like to undo changes across multiple buffers in a coherent way -- no idea haha.